alibuild_helpers/build.py

```diff
@@ -363,6 +367,72 @@ def createDistLinks(spec, specs, args, repoType, requiresType):
                  t=target)
     execute(cmd)
 
+
+def storeHashes(package, specs, isDevelPkg, considerRelocation):
+  """Calculate various hashes for package, and store them in specs[package].
+
+  Assumes that all dependencies of the package already have a definitive hash.
+  """
+  spec = specs[package]
+
+  if "remote_revision_hash" in spec and "local_revision_hash" in spec:
+    # We've already calculated these hashes before, so no need to do it again.
+    # This also works around a bug, where after the first hash calculation,
+    # some attributes of spec are changed (e.g. append_path and prepend_path
+    # entries are turned from strings into lists), which changes the hash on
+    # subsequent calculations.
+    return
+
```

Comment:
Sorry, but this make it really difficult to understand what changed. Can we have the refactoring separate from the logic change?